[mlir][ROCDL] Adds wmma scaled intrinsics for gfx1250

Muzammiluddin-Syed-ECE opened the pull request.

kuhar commented on mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:
```diff
@@ -627,6 +627,8 @@ def ROCDL_wmma_f16_16x16x128_fp8_bf8 : ROCDL_Wmma_IntrOp<"wmma.f16.16x16x128.fp8
 def ROCDL_wmma_f16_16x16x128_bf8_bf8 : ROCDL_Wmma_IntrOp<"wmma.f16.16x16x128.bf8_bf8", [0]>;
 def ROCDL_wmma_i32_16x16x64_iu8 : ROCDL_Wmma_IntrOp<"wmma.i32.16x16x64.iu8", [1]>;
+def ROCDL_wmma_scale_f32_16x16x128_f8f6f4   : ROCDL_Wmma_IntrOp<"wmma.scale.f32.16x16x128.f8f6f4", [1,3]>;
+def ROCDL_wmma_scale16_f32_16x16x128_f8f6f4   : ROCDL_Wmma_IntrOp<"wmma.scale16.f32.16x16x128.f8f6f4", [1,3]>;
```
What about the variants with M=32?

kuhar reviewed (approved):
LGTM but let's wait for @krzysz00 to confirm

krzysz00 reviewed (commented):
I think this might be our cue to go fix up WMMA to actually encode argument types properly. Because to quote `AMDGPUWmmaScaleIntrinsicModsC`,

```
ImmArg<ArgIndex<0>>, ImmArg<ArgIndex<2>>, ImmArg<ArgIndex<4>>, ImmArg<ArgIndex<6>>,
     ImmArg<ArgIndex<7>>, ImmArg<ArgIndex<9>>, ImmArg<ArgIndex<10>>, ImmArg<ArgIndex<12>>, ImmArg<ArgIndex<13>>,
```

That's a lot of things that should be attributes in the MLIR

Muzammiluddin-Syed-ECE committed:
[ROCDL] Add scaled WMMA intrinsics for gfx1250

This patch adds support for scaled WMMA intrinsics available on gfx1250+.

Key changes:
- Created two classes for intrinsic variants ROCDL_WMMA_Scale_IntrOp and ROCDL_WMMA_Scale_F4_IntrOp.
- Tests in rocdl.mlir updated.

Intrinsics added:
- rocdl.wmma.scale.f32.16x16x128.f8f6f4
- rocdl.wmma.scale16.f32.16x16x128.f8f6f4
- rocdl.wmma.scale.f32.32x16x128.f4
- rocdl.wmma.scale16.f32.32x16x128.f4

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>

Muzammiluddin-Syed-ECE committed:
renaming attributes

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>

krzysz00 reviewed (commented):
Could you update tests/Dialect/LLVMIR/rocdl.mlir too? It could use a printer-parser test

Muzammiluddin-Syed-ECE committed:
addressing pr comments

Signed-off-by: Muzammiluddin Syed <muzasyed@amd.com>

krzysz00 reviewed (approved):
Approved, thanks
